src/mds/SnapRealm.cc

```diff
@@ -339,15 +341,21 @@ void SnapRealm::split_at(SnapRealm *child)
   }
 
   // split inodes_with_caps
-  for (auto p = inodes_with_caps.begin(); !p.end(); ) {
-    CInode *in = *p;
-    ++p;
-    // does inode fall within the child realm?
-    if (child->inode->is_ancestor_of(in)) {
-      dout(20) << " child gets " << *in << dendl;
-      in->move_to_realm(child);
-    } else {
-      dout(20) << "    keeping " << *in << dendl;
+  if (child->inode->dirfragtree_empty()) {
```

Comment:
So this doesn't work because directories may be "empty" because there are no HEAD dentries but some snapshotted dentries exist, i.e. `CDir::get_num_snap_items() > 0`. So there will be some inodes with caps, potentially, that refer to the snapshotted inodes associated with those dentries.

The solution, I believe, is to create a hierarchical data structure part of `CDir` which allows tracking child inodes with caps or something that allows us to optimize the common cases (small directories with few or 0 snapshotted dentries).

Reply:
OK, so the "parent" snaprealm tracks all the inodes under its tree which have caps and the root directory always has a snaprealm which implies that every inode under it used by a client is in the inodes_with_caps list for the root directory. And then when a child snapshot is created, the split_at iterates through the entire (probably millions) entries and "moves" those inodes which now fall under the child snap realm.

Reply:
> So this doesn't work because directories may be "empty" because there are no HEAD dentries but some snapshotted dentries exist, i.e. `CDir::get_num_snap_items() > 0`. So there will be some inodes with caps, potentially, that refer to the snapshotted inodes associated with those dentries.
> 
> The solution, I believe, is to create a hierarchical data structure part of `CDir` which allows tracking child inodes with caps or something that allows us to optimize the common cases (small directories with few or 0 snapshotted dentries).

If the child inodes are tracked at their respective parent directory level, wouldn't there exist a similar problem around when accumulating inodes for an "upper" directory level snap realm?